`TestDiagnoseSSHConnection` flakiness

## Failure

#### Link(s) to logs

- https://console.cloud.google.com/cloud-build/builds/12a4f182-732d-4e25-806d-6a8530324565;step=0?project=ci-account

#### Relevant snippet

Observed on branch/v10:

```
    apiserver_test.go:4094: principal doesnt exist in target host
    apiserver_test.go:4095: failed false
    apiserver_test.go:4101: 0 status='SUCCESS' type='RBAC_NODE' details='You have access to the Node.' error=''
    apiserver_test.go:4101: 1 status='SUCCESS' type='CONNECTIVITY' details='Node is alive and reachable.' error=''
    apiserver_test.go:4101: 2 status='SUCCESS' type='RBAC_PRINCIPAL' details='The requested principal is allowed.' error=''
    apiserver_test.go:4101: 3 status='FAILED' type='CONNECTIVITY' details='Failed to connect to the Node. Ensure teleport service is running using "systemctl status teleport".' error='connection canceled'
    apiserver_test.go:4124: 
        	Error Trace:	apiserver_test.go:4124
        	Error:      	Should be true
        	Test:       	TestDiagnoseSSHConnection/principal_doesnt_exist_in_target_host
        	Messages:   	{Type:NODE_PRINCIPAL Status:FAILED Details:Invalid user. Please ensure the principal "nonvalidlinuxuser" is a valid Linux login in the target node. Output from Node: Failed to launch: user: unknown user nonvalidlinuxuser. Error:Process exited with status 255 XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
```

https://console.cloud.google.com/cloud-build/builds/3278dced-e8ba-4029-a2b2-d7b15f94db6a?project=ci-account

```
    --- FAIL: TestDiagnoseSSHConnection/success (5.54s)
        	Test:       	TestDiagnoseSSHConnection/success
        	            	actual  : false
        	            	expected: true
        	Error:      	Not equal: 
        	Error Trace:	/workspace/lib/web/apiserver_test.go:4464
    apiserver_test.go:4464: 
    apiserver_test.go:4461: 3 status='FAILED' type='CONNECTIVITY' details='Failed to connect to the Node. Ensure teleport service is running using "systemctl status teleport".' error='connection canceled'
    apiserver_test.go:4461: 2 status='SUCCESS' type='RBAC_PRINCIPAL' details='The requested principal is allowed.' error=''
    apiserver_test.go:4461: 1 status='SUCCESS' type='CONNECTIVITY' details='Node is alive and reachable.' error=''
    apiserver_test.go:4461: 0 status='SUCCESS' type='RBAC_NODE' details='You have access to the Node.' error=''
    apiserver_test.go:4455: failed false
    apiserver_test.go:4454: success
```

Got another failure here: https://github.com/gravitational/teleport/actions/runs/4631724318/jobs/8194939954.

```
    apiserver_test.go:5759: 
        	Error Trace:	/__w/teleport/teleport/lib/web/apiserver_test.go:5759
        	Error:      	Should be true
        	Test:       	TestDiagnoseSSHConnection/no_access_to_node
        	Messages:   	expected trace {RBAC_NODE FAILED You are not authorized to access this node. Ensure your role grants access by adding it to the 'node_labels' property. user userwithoutaccess@localhost is not authorized to login as root@localhost: access to node denied {} [] 0} was not found
--- FAIL: TestDiagnoseSSHConnection/no_access_to_node (6.84s)
```

@marcoandredinis Can you take a look please?

The test expected failure, but got timeout instead

https://github.com/gravitational/teleport/actions/runs/5142163720/jobs/9255550036

```
    apiserver_test.go:5844: no access to node
    apiserver_test.go:5845: failed false
    apiserver_test.go:5851: 0 status='FAILED' type='UNKNOWN_ERROR' details='Unknown error. ' error='context deadline exceeded'
    apiserver_test.go:5874: 
        	Error Trace:	/__w/teleport/teleport/lib/web/apiserver_test.go:5874
        	Error:      	Should be true
        	Test:       	TestDiagnoseSSHConnection/no_access_to_node
        	Messages:   	expected trace {RBAC_NODE FAILED You are not authorized to access this node. Ensure your role grants access by adding it to the 'node_labels' property. user userwithoutaccess@localhost is not authorized to login as root@localhost: access to node denied {} [] 0} was not found
--- FAIL: TestDiagnoseSSHConnection/no_access_to_node (6.83s)
```

--- a/build.assets/tooling/cmd/difftest/main.go
+++ b/build.assets/tooling/cmd/difftest/main.go
@@ -60,6 +60,10 @@ var (
 		// TestProxySSH and TestList takes around 10-15s to run, largely due to the 7-10 seconds it takes to create a
 		// tsh test suite. This prevents it from ever completing the 100 runs successfully.
 		"TestProxySSH", "TestList", "TestForwardingTraces", "TestExportingTraces",
+
+		// TestDiagnoseSSHConnection takes around 15s to run.
+		// When running 100x it exceeds the 600s defined to run the tests.
+		"TestDiagnoseSSHConnection",
 	}
 )
 

--- a/lib/web/apiserver_test.go
+++ b/lib/web/apiserver_test.go
@@ -5834,6 +5834,17 @@ func TestDiagnoseSSHConnection(t *testing.T) {
 	env := newWebPack(t, 1)
 	nodeName := env.node.GetInfo().GetHostname()
 
+	// Wait for node to show up
+	require.Eventually(t, func() bool {
+		_, err := env.server.Auth().GetNode(ctx, apidefaults.Namespace, nodeName)
+		if trace.IsNotFound(err) {
+			return false
+		}
+		require.NoError(t, err)
+
+		return true
+	}, 5*time.Second, 250*time.Millisecond)
+
 	for _, tt := range []struct {
 		name            string
 		teleportUser    string
@@ -5980,9 +5991,6 @@ func TestDiagnoseSSHConnection(t *testing.T) {
 				ResourceKind: types.KindNode,
 				ResourceName: tt.resourceName,
 				SSHPrincipal: tt.nodeUser,
-				// Default is 30 seconds but since tests run locally, we can reduce this value to also improve test responsiveness
-				// A value too low (eg 1s) will introduce test flakiness on some systems.
-				DialTimeout: 5 * time.Second,
 			})
 			require.NoError(t, err)
 			require.Equal(t, http.StatusOK, resp.Code())
@@ -6052,9 +6060,7 @@ func TestDiagnoseSSHConnection(t *testing.T) {
 		ResourceKind: types.KindNode,
 		ResourceName: nodeName,
 		SSHPrincipal: osUsername,
-		// Default is 30 seconds but since tests run locally, we can reduce this value to also improve test responsiveness
-		DialTimeout: 5 * time.Second,
-		MFAResponse: client.MFAChallengeResponse{TOTPCode: totpCode},
+		MFAResponse:  client.MFAChallengeResponse{TOTPCode: totpCode},
 	})
 	require.NoError(t, err)
 	require.Equal(t, http.StatusOK, resp.Code())

--- a/lib/web/apiserver_test.go
+++ b/lib/web/apiserver_test.go
@@ -59,6 +59,7 @@ import (
 	"github.com/pquerna/otp/totp"
 	"github.com/sashabaranov/go-openai"
 	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	commonv1 "go.opentelemetry.io/proto/otlp/common/v1"
 	resourcev1 "go.opentelemetry.io/proto/otlp/resource/v1"
@@ -5840,8 +5841,7 @@ func TestDiagnoseSSHConnection(t *testing.T) {
 		if trace.IsNotFound(err) {
 			return false
 		}
-		require.NoError(t, err)
-
+		assert.NoError(t, err, "GetNode returned an unexpected error")
 		return true
 	}, 5*time.Second, 250*time.Millisecond)
 

